[API] Finished Implementation of User Achievement Creation endpoint

abrodriguezalman opened the pull request:
In partnership with the Achievements team, this issue aims to develop an API endpoint to allow users to assign a particular ACHIEVEMENT from a given GAME to a specified USER. This PR:
1. Creates a serializer for the UserAchievement model created by the Achievements team
2. Implements a new function UserAchievementCreateView() under api/views.py with a corresponding URL path "api/games/game_id/achievements/achievement_id/assign/".

These two modifications allow users to assign a particular achievement for a given game to a specified user.

Commit:
[API] Finished Implementation of User Achievement Creation endpoint

abrodriguezalman committed:
Merge branch 'dev' into api/achievement-user-creation-endpoint

MuyanXie committed:
Merge branch 'dev' into api/achievement-user-creation-endpoint

MuyanXie commented:
The endpoint is not working for me. When I use the date toggle, it shows incorrect format. Please make sure that the users can simply use the HTML form and choose a time there -- maybe change the serializer/view so that it's able to translate time in different formats, or change the frontend view/style for the time choice

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/ed4f6d2a-927b-4d42-9eb5-03e16e868a1d" />

MuyanXie reviewed (changes requested):
Please see comments above

Commit:
[API] Added success message to be returned after successful POST request in UserAchievementCreateView(). Removed unnecessary game parameter.

abrodriguezalman commented:
> The endpoint is not working for me. When I use the date toggle, it shows incorrect format. Please make sure that the users can simply use the HTML form and choose a time there -- maybe change the serializer/view so that it's able to translate time in different formats, or change the frontend view/style for the time choice
> 
> <img alt="image" width="1512" src="https://private-user-images.githubusercontent.com/72187977/443305008-ed4f6d2a-927b-4d42-9eb5-03e16e868a1d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDcxNTQwOTMsIm5iZiI6MTc0NzE1Mzc5MywicGF0aCI6Ii83MjE4Nzk3Ny80NDMzMDUwMDgtZWQ0ZjZkMmEtOTI3Yi00ZDQyLTllYjUtMDNlMTZlODY4YTFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTEzVDE2Mjk1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIyYjdiODkyZjQ1YmVlN2U5YTAyNDg0ODEyYzZjOTA1OTMwN2U5NjkzMDlhODE5ZjA1NTIzYzA4ZTU2MTlmY2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XC4LyUrJuOE2ZjXdmaqulc4Xfij-XPwtWMEVyNIA2pY">

@MuyanXie That is a bit weird; the HTML form seems to be working correctly on my end, as seen below. Were you able to click on the date/time values to change them?
<img width="959" alt="Screenshot 2025-05-13 at 12 14 37 PM" src="https://github.com/user-attachments/assets/a707a87f-eb6b-40c7-8f15-88d60e745b34" />

chrislowzhengxi reviewed (approved):
I tested the UserAchievementCreateView manually using the POST endpoint /api/games/1/achievements/1/assign/ with:

`{
  "user": 1,
  "date_earned": "2025-05-22T16:00:00Z",
  "progress": 5,
  "pinned": false
}`

and received a 201 Created response with the correct data.
<img width="1239" alt="Screenshot 2025-05-13 at 12 45 33 PM" src="https://github.com/user-attachments/assets/1a2b6d4d-4b3f-4fec-bcaf-a594ce7c5956" />


I also confirmed that the UserAchievementCreateView works correctly through the HTML form interface. I was able to select a user, choose a datetime using the date picker, and input progress, as shown. 

<img width="1257" alt="Screenshot 2025-05-13 at 12 47 38 PM" src="https://github.com/user-attachments/assets/5fec119b-6e2b-4bb9-b023-4e174f996a87" />


Minor: Currently, reassigning an existing achievement to the same user causes a UNIQUE constraint failed IntegrityError. Might be worth catching this at the view or serializer level and returning a 400 Bad Request with a clearer message for clients.

MuyanXie commented:
@chrislowzhengxi @abrodriguezalman thanks! I'm able to refresh and get the data error fixed. I encountered the following error though:

TypeError at /api/games/1/achievements/1/assign/
Got a `TypeError` when calling `UserAchievement.objects.create()`. This may be because you have a writable field on the serializer class that is not a valid argument to `UserAchievement.objects.create()`. You may need to make the field read-only, or override the UserAchievementSerializer.create() method to handle this correctly.
Original exception was:
 Traceback (most recent call last):
  File "/Users/muyanxiesam/Documents/school_work/CMSC 22000/chigame/venv/lib/python3.13/site-packages/rest_framework/serializers.py", line 962, in create
    instance = ModelClass._default_manager.create(**validated_data)
  File "/Users/muyanxiesam/Documents/school_work/CMSC 22000/chigame/venv/lib/python3.13/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/Users/muyanxiesam/Documents/school_work/CMSC 22000/chigame/venv/lib/python3.13/site-packages/django/db/models/query.py", line 656, in create
    obj = self.model(**kwargs)
  File "/Users/muyanxiesam/Documents/school_work/CMSC 22000/chigame/venv/lib/python3.13/site-packages/django/db/models/base.py", line 567, in __init__
    raise TypeError(
    ...<2 lines>...
    )
TypeError: UserAchievement() got unexpected keyword arguments: 'game_id'

What I did: I navigated to /api/games/1/achievements/1/assign/; I can verify I have a game and a achieve in db as verified by the admin dashboard. I post the common payload with user being the admin

Commit:
[API] Modified UserAchievementCreateView() to send 400 error message to meet the unique constraint between achievement and user for a game

abrodriguezalman commented:
I have updated the UserAchievementCreateView() function so that a 400 error response is sent whenever a user tries to make a UserAchievement object with an achievement that has already been assign to the user. I also added a small error message to indicate the source of the problem to users. This commit should also fix the error @MuyanXie faced due to a misplaced "game_id" variable.

abrodriguezalman committed:
Merge branch 'dev' into api/achievement-user-creation-endpoint

chrislowzhengxi commented:
LGTM! I was able to verify that when I post the same achievement twice, I get an error as shown. 
<img width="1285" alt="Screenshot 2025-05-14 at 8 36 56 PM" src="https://github.com/user-attachments/assets/f2380439-323e-4a76-980a-03f6de2d24ec" />

chrislowzhengxi reviewed (approved):
LGTM! I was able to verify that when I post the same achievement twice, I get an error as shown.

`HTTP 400 Bad Request
Allow: POST, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "error": "This achievement already exists for user '[chrislow@example.com](mailto:chrislow@example.com)'"
}`

MuyanXie reviewed (approved):
It looks good to me on my end now. Please resolve the merge conflict so that I can merge

abrodriguezalman committed:
Merge branch 'dev' into api/achievement-user-creation-endpoint

Commit:
[API] addressed merge conflicts

Commit:
[API] addressed merge conflicts

abrodriguezalman commented:
Concluding statement
In order to address this issue, the following changes were implemented to the codebase:

Creates a serializer for the UserAchievement model created by the Achievements team
Implements a new function UserAchievementCreateView() under api/views.py with a corresponding URL path "api/games/game_id/achievements/achievement_id/assign/".

These two modifications allow users to assign a particular achievement for a given game to a specified user.

MuyanXie commented:
Issue Score: excellent

Comments:
Nice job! The only thing to pay attention to is that you might want to thoroughly test before requesting final review, but still it's a good PR overall!

<!-- {"score": "excellent", "minor": false} -->
